host/xtest/aes_perf.c

```diff
@@ -0,0 +1,474 @@
+	check_res(res, "TEEC_InvokeCommand");
+}
+
+static void do_warmup(int warmup)
+{
+	struct timespec t0, t;
+	int i;
+
+	get_current_time(&t0);
+	do {
+		for (i = 0; i < 100000; i++)
+			;
+		get_current_time(&t);
+	} while (timespec_diff_ns(&t0, &t) < (uint64_t)warmup * 1000000000);
+}
+
+static const char *yesno(int v)
+{
+	return (v ? "yes" : "no");
+}
+
+static double mb_per_sec(size_t size, double usec)
+{
+	return (1000000000/usec)*((double)size/(1024*1024));
+}
+
+/* Encryption test: buffer of tsize byte. Run test n times. */
+void aes_perf_run_test(int mode, int keysize, int decrypt, size_t size,
+				unsigned int n, unsigned int l, int random_in,
+				int in_place, int warmup, int verbosity)
+{
+	uint64_t t;
+	struct statistics stats;
+	struct timespec ts;
+	TEEC_Operation op;
+	int n0 = n;
+
+	vverbose("aes-perf version %s\n", TO_STR(VERSION));
+	if (clock_getres(CLOCK_MONOTONIC, &ts) < 0) {
+		perror("clock_getres");
+		return;
+	}
+	vverbose("Clock resolution is %lu ns\n", ts.tv_sec*1000000000 +
+		ts.tv_nsec);
+
+	open_ta();
+	prepare_key(decrypt, keysize, mode);
+
+	memset(&stats, 0, sizeof(stats));
+
+	alloc_shm(size, in_place);
+
+	if (!random_in)
+		memset(in_shm.buffer, 0, size);
+
+	memset(&op, 0, sizeof(op));
+	/* Using INOUT to handle the case in_place == 1 */
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_PARTIAL_INOUT,
+					 TEEC_MEMREF_PARTIAL_INOUT,
+					 TEEC_VALUE_INPUT, TEEC_NONE);
+	op.params[0].memref.parent = &in_shm;
+	op.params[0].memref.offset = 0;
+	op.params[0].memref.size = size;
+	op.params[1].memref.parent = in_place ? &in_shm : &out_shm;
+	op.params[1].memref.offset = 0;
+	op.params[1].memref.size = size;
+	op.params[2].value.a = l;
+
+	verbose("Starting test: %s, %scrypt, keysize=%u bits, size=%zu bytes, ",
+		mode_str(mode), (decrypt ? "de" : "en"), keysize, size);
+	verbose("random=%s, ", yesno(random_in));
+	verbose("in place=%s, ", yesno(in_place));
+	verbose("inner loops=%u, loops=%u, warm-up=%u s\n", l, n, warmup);
+
+	if (warmup)
+		do_warmup(warmup);
+
+	while (n-- > 0) {
+		t = run_test_once(in_shm.buffer, size, &op, random_in);
+		update_stats(&stats, t);
+		if (n % (n0/10) == 0)
+			vverbose("#");
+	}
+	vverbose("\n");
+	printf("min=%gμs max=%gμs mean=%gμs stddev=%gμs (%gMiB/s)\n",
+	       stats.min/1000, stats.max/1000, stats.m/1000,
+	       stddev(&stats)/1000, mb_per_sec(size, stats.m));
+	free_shm();
+}
+
+#define NEXT_ARG(i) \
+	do { \
+		if (++i == argc) { \
+			fprintf(stderr, "%s: %s: missing argument\n", \
+				argv[0], argv[i-1]); \
+			return 1; \
+		} \
+	} while (0);
+
+int aes_perf_runner_cmd_parser(int argc, char *argv[])
+{
+	int i;
+
+	/*
+	* Command line parameters
+	*/
+
+	size_t size = 1024;	/* Buffer size (-s) */
+	unsigned int n = 5000;	/* Number of measurements (-n) */
+	unsigned int l = 1;	/* Inner loops (-l) */
+	int verbosity = 0;	/* Verbosity (-v) */
+	int decrypt = 0;		/* Encrypt by default, -d to decrypt */
+	int keysize = 128;	/* AES key size (-k) */
```

Comment:
That's totally useless IMO. `TA_AES_ECB` etc. are defined because the client and the secure side have to agree on a way to define the AES modes. I could as well have chosen to transmit a string such as "ECB" or "ecb", but I found it slightly simpler to use numbers. There is no such need for the key size obviously...